src/c/urweb.c

```diff
@@ -2347,7 +2346,21 @@ uw_unit uw_Basis_htmlifySpecialChar_w(uw_context ctx, uw_Basis_char ch) {
   int len;
 
   uw_check(ctx, INTS_MAX+3);
-  len = sprintf(ctx->page.front, "&#%u;", n);
+
+  if(uw_Basis_isprint(ctx, ch)) {
+
+    UChar32 ins[1] = { ch };
+    char buf[5];
+    int32_t len_written = 0;
+    UErrorCode err = U_ZERO_ERROR;
+
+    u_strToUTF8(buf, 5, &len_written, ins, 1, &err);
```

Comment:
Minor nitpicking, I think the snippet

```C
UChar32 ins[1] = { ch };
u_strToUTF8(buf, 5, &len_written, ins, 1, &err);
```
can be replaced with

```C
u_strToUTF8(buf, 5, &len_written, (const UChar*)&ch, 1, &err);
```

(without the cast maybe will not compile due to warning level of compiler)

Reply:
Also I think you can leave out `buf` if you write directly to `ctx->page.front`:

```C
u_strToUTF8(ctx->page.front, 5, &len_written, (const UChar*)&ch, 1, &err);
```